### ISSUE TITLE
Autoplay for casted Videos

### ISSUE THREAD

**Issue by francescosoft** [enen92/script.tubecast#15]
Thanks for this very nice add-on.
Is it possible to implement autoplay for the Casted Videos like on Chromecast ?

**Comment by enen92** [comment]
If I recall correctly this is implemented by the youtube web application on the chromecast and not really on the mobile app. So, there is no info sent from the phone to the cast device regarding which is the next video to be played. I'll look closely

**Comment by enen92** [comment]
With the developments being done in kodi (which is almost ready to support a web-browser and web-based addons) this might well be a possibility in the near future.

**Comment by silelmot** [comment]
this would be really great. this means it would be possible to let youtube.tv be run on kodi just from an app? where can we read more about it?

**Comment by enen92** [comment]
This means I'll add an option to just display the YouTube official app and let it take care of the websocket connection. The only thing the add-on will do with that option enabled is the network discovery step.

**Comment by cyraid** [comment]
@silemont youtube.tv will no longer work on any unofficial devices after October 2nd.. (test it yourself now in a browser and see the warning).

@enen92 might have to keep it like this for the future since Google is taking down youtube.tv (leanback). Unless you mean the official app that's on the Chromecast?

**Comment by silelmot** [comment]
@cyraid i already saw this. this is a big step down in convinience.
tubecast would fill this hole, but as long as querying doesn't work this is no good replacment.
i am sorry, i can not code myself good enough to be any help

**Comment by siku2** [comment]
I realise that this is an ancient issue and I would feel bad for necroposting if it weren't still relevant.

I find it kind of weird that TubeCast currently always *enables* autoplay even though it doesn't support it (See the [`_set_disabled`](https://github.com/enen92/script.tubecast/blob/e782390/resources/lib/tubecast/youtube/app.py#L368) method which is called every time a client connects). Based on the method name I think it's meant to disable it but I might be missing something.

As has been said before, auto-play almost certainly needs to be handled by the host (i.e. TubeCast). But good news, the [Kodi YouTube plugin](https://github.com/jdf76/plugin.video.youtube) (TubeCast uses this to play the videos) can be used as a module. This means we can use it to find the recommended videos and play them if auto-play is activated.
I don't know how TubeCast would then add it to the playlist though. Currently the app has full control over the playlist but I'm sure there's way.

There is something I need to know though and I would greatly appreciate if someone could test this for me.
Does auto-play on the official TV always select the first recommendation of the last video in the playlist or does it take the next video from somewhere else?

**Comment by movedoa** [comment]
@siku2 I tried it with a few videos on the smart tv app and it seems autoplay always selects the first recommended video.

**Comment by siku2** [comment]
Thank you, @movedoa! This is great news because the YouTube plugin already has this functionality implemented.

**Comment by movedoa** [comment]
..ok wow i tried like 20 videos. Just after i posted the comment here i got a different video. It is still in the recommended videos but it was the 14th in the list.

**Comment by siku2** [comment]
Huh, that's odd. I have a suspicion though. Do you think i'st possible that YouTube somehow picks a video based on ALL videos that are currently in the queue, no just the last one?

I know that they already have this functionality for playlists so I don't think it's unreasonable to think that they would employ the same strategy here.

**Comment by movedoa** [comment]
I had a video now where the autoplay video was not in the recommendations at all. I cleared my playlist and restarted the video and got the same recommendation again. It seems there is something more going on here. The next video was from a channel i subscribe to though. So i guess the youtube algorythm just gets a video based on x factors.

Anyway, selecting the first video would imho be a good solution for tubecast. I don't think anyone depends on the "random" next video. The first recommendation should be a fitting next video most of the time.

**Comment by siku2** [comment]
> Anyway, selecting the first video would imho be a good solution for tubecast.

I agree, this will definitely do for a start.

The only thing I'm worried about is that YouTube recommendations have a tendency to repeat themselves after a while (At least this was the case the last time I used them for something similar).
We could however actively avoid picking videos that are already in the playlist...

**Comment by movedoa** [comment]
Pick the first video, if it is in the playlist already pick the first one down the line that isn't. If all videos are in the playlist already fall back to the first one again.

Even the official app sometimes replays a video so i don't think this is a real problem. Worst case is that the user has to select something manually.

**Comment by enen92** [comment]
Carefull as this addon doesn't require that the youtube addon has a login defined. If such feature is enabled the necessary checks for the presence of credentials also need to be also addressed

**Comment by siku2** [comment]
Good point but I think the "related" videos can be retrieved without any authorization.
Will definitely look out for this though, thanks.
